### PR TITLE
[DRAFT] Blackhole Mistral-Small-4-119B-2603 Bringup

### PR DESCRIPTION
**Pull request by ign-ssunil**
### Problem description

Adding **Mistral-Small-4-119B-2603** to `models/experimental/mistral_small_4_119b`: a 119B-parameter (6.5B active/token) **multimodal Mixture-of-Experts** model — Pixtral vision tower + MLA/MoE language decoder — running end to end on Tenstorrent Blackhole hardware with **chunked prefill** and a **2-command-queue traced decode** pipeline.

### Platforms

| Device | Status | Notes |
|--------|--------|-------|
| BH Loudbox (Blackhole) / P150×8 | Supported | `MeshShape(1, 8)` — 8 chips, tensor-parallel; 2CQ traced decode on by default |

This port targets the **1×8 BlackHole P150 mesh exclusively**. Set `MESH_DEVICE=P150x8` before running (see `tests/mesh_param.py`).

### Introduction

[Mistral-Small-4-119B-2603](https://huggingface.co/mistralai/Mistral-Small-4-119B-2603) is Mistral AI's 119B multimodal MoE model (128 experts, 4 active + 1 shared; 6.5B active params/token), pairing a **Pixtral vision tower (24 layers)** with a **36-layer MLA + MoE text decoder** (Multi-head Latent Attention with a compressed latent KV cache). This port runs the full vision → projector → language pipeline on a Blackhole mesh via TTNN.

Loading is **unified**: vision + projector + text weights lazy-load together on the first `encode_image` call. Weights are stored in **mixed precision** — most linear projections in `bfloat4_b`/`bfloat8_b` (attention `q_a`/`kv_a`/`o_proj` and `lm_head` in `bfloat4_b`; `q_b`, MoE experts, and the vision tower in `bfloat8_b`), while layernorms, the `wkv_b` absorption matrices, embeddings, and the KV cache stay in `bfloat16`; activations run in `bfloat16`. Decode is trace-captured and replayed over 2 command queues (`decode_next_token_2cq`); prefill is chunked and **not** trace-captured (it builds a fresh device input and reads the next-token id back to host each call).

**Full-model PCC:** ~**0.855** on the unified multimodal path (consistent with the aggressive `bfloat4_b`/`bfloat8_b` weight quantization above). PCC harnesses are provided for the vision tower, projector, prefill logits, decode step, and the unified multimodal path.

**Text decoder prefill PCC (full 36 layers, BH Loudbox / P150×8):** Overall flattened logits PCC = **0.933**, mean per-position PCC = **0.911** (prompt: `"The capital of France is"`, 5 tokens, vs HF reference). Position 0 PCC = 0.717 (expected low — degenerate causal context); positions 1–4 all ≥ 0.949. Greedy token match: 4/5 (only position 0 differs). Run with `pytest tests/test_text_prefill_pcc.py` (defaults to all 36 layers; `@pytest.mark.slow`).

**Text decoder decode PCC (full 36 layers, BH Loudbox / P150×8):** Two checks at position 4 (same 5-token prompt): decode-vs-HF logits PCC = **0.961** (ground truth, floor 0.90) and decode-vs-prefill self-consistency PCC = **0.992** (floor 0.70). Greedy token agrees three ways — HF = prefill = decode = `6993`. The self-consistency check isolates whether decode correctly reads/uses the KV cache prefill wrote, free of the bfloat16 kernel-shape artifact; the HF check anchors absolute correctness and folds that artifact in on top of quantization loss. Run with `pytest tests/test_text_decode_pcc.py` (defaults to all 36 layers; `@pytest.mark.slow`).

**Text prefill PCC — ISL sweep (prefill logits vs HF, full 36 layers, BH Loudbox / P150×8):** `test_text_prefill_pcc.py` sweeps input sequence length — `base` (the 5-token prompt above) plus coherent-English-filler prompts at 128 / 512 / 2K / 4K / 16K tokens (select with `-k`). Both the TTNN prefill and the HF reference are chunked (HF via `MISTRAL4_PREFILL_HF_CHUNK`, default 2048, DynamicCache → O(chunk × past) memory), so the comparison scales beyond a single short prompt:

| ISL (tokens) | Prefill flattened PCC | Greedy token match |
|-------------:|----------------------:|-------------------:|
| 5 (base) | 0.933 | 4/5 |
| 127 | 0.951 | 115/127 |
| 505 | 0.927 | 413/505 |
| 2,017 | 0.914 | 1,499/2,017 |
| 4,033 | 0.913 | 2,676/4,033 |
| 16,045 | 0.905 | 7,885/16,045 |

Flattened PCC stays **≥ 0.90 through 16K** and asymptotes ~0.91; the greedy-match rate falls (91% → 49%) purely because longer filler has more high-entropy positions where argmax is near-tied — the flattened PCC is the gate (per-position min dips at those positions by design). Decode-vs-HF was also spot-checked at depth: **0.961** at the 5-token base and **0.887** at 16K context (single-position, so noisier than the averaged prefill PCC, and the 16K filler tail is a high-entropy EOS position).

**Maximum context length:** The HF checkpoint advertises up to **256K** tokens (RoPE θ=1e6). On Blackhole Loudbox (1×8), correctness is **PCC-verified against the HF reference up to ~16K tokens end-to-end — vision + language**. The parametrized `test_multimodal_pcc_unified.py` sweeps fixed `(num_text_layers, num_vision_layers, max_text_tokens)` points — ids `L1V1`, `L36V24`, `L36V24_4096`, `L36V24_16384`, `L36V24_65536`, `L36V24_131072`, `L36V24_262144` — selected with `-k` (e.g. `-k L36V24_16384`). For each point it pads the prompt with coherent English filler up to `max_text_tokens` (image tokens added on top), runs the HF reference chunked via `MISTRAL4_MM_PREFILL_CHUNK` (default 2048, bounding attention memory to O(chunk × past)), and compares per-position + flattened logits PCC. PCC ≈ **0.839 / 0.842** at 4K / 16K incl. image tokens. **16K is the highest context that completes PCC verification** — the `L36V24_65536`, `L36V24_131072`, and `L36V24_262144` points are kept (marked `@pytest.mark.slow`) as a safeguard, but in practice they do not finish on a CPU host: a 64K run was killed after **> 1 h without completing**, bottlenecked on the wall-clock time of the chunked HF CPU reference forward (chunking already bounds its memory, so this is a compute-time ceiling, not RAM). On-device prefill + decode themselves scale much further (see the perf table, which uses synthetic inputs and no HF reference), but are not PCC-verified past 16K.

### Prerequisites

1. Cloned [tt-metal repository](https://github.com/tenstorrent/tt-metal)
2. Installed [TT-Metalium / TT-NN](https://github.com/tenstorrent/tt-metal/blob/main/INSTALLING.md)
3. Mistral4 support is not in a released `transformers` yet — install from source:
   ```bash
   uv pip install git+https://github.com/huggingface/transformers.git
   ```
4. Access to the gated HF model:
   ```bash
   huggingface-cli login
   ```

### Environment Setup

```bash
source python_env/bin/activate
export TT_METAL_HOME=$(pwd)
export PYTHONPATH="${TT_METAL_HOME}"

# Optional: cache loaded weights for faster subsequent runs
export MISTRAL4_WEIGHT_CACHE_DIR=<weight_cache_dir>
```

#### PCC tests

The text prefill and decode PCC tests run the **full 36-layer** decoder by default and are marked
`@pytest.mark.slow` (override layer count with `MISTRAL4_PREFILL_N_LAYERS` / `MISTRAL4_DECODE_N_LAYERS`
for fast local iteration). The remaining harnesses stay gated behind per-test env flags.

```bash
pytest models/experimental/mistral_small_4_119b/tests/test_text_prefill_pcc.py   # full 36 layers
pytest models/experimental/mistral_small_4_119b/tests/test_text_decode_pcc.py    # full 36 layers

MISTRAL4_VISION_PCC=1  pytest models/experimental/mistral_small_4_119b/tests/test_vision_tower_pcc.py
MISTRAL4_MMP_PCC=1     pytest models/experimental/mistral_small_4_119b/tests/test_multimodal_projector_pcc.py
MISTRAL4_MM_PCC=1      pytest models/experimental/mistral_small_4_119b/tests/test_multimodal_pcc_unified.py
```

#### Demo

Multimodal generation via [`demo/demo_multimodal.py`](demo/demo_multimodal.py). The default image and prompt run the full 36-text + 24-vision-layer model:

```bash
export MESH_DEVICE=P150x8
python models/experimental/mistral_small_4_119b/demo/demo_multimodal.py \
    --image-max-side 512 \
    --n-text-layers 36 \
    --n-vision-layers 24 \
    --max-new-tokens 512
```

Override the input with `--image <path-or-url>` and `--prompt "<text>"`.

### Performance (Blackhole Loudbox, 2CQ traced decode)

Measured with `pytest models/experimental/mistral_small_4_119b/tests/perf/test_e2e_performant.py -m models_performance_bare_metal -k L36V24`.

| ISL (tokens) | TTFT (ms) | Prefill tok/s | Decode tok/s/user |
|-------------:|----------:|--------------:|------------------:|
| 128 | 475.8 | 301.5 | 16.99 |
| 4096 | 3,392.5 | 1,225.5 | 16.32 |
| 16384 | 12,609.6 | 1,305.2 | 15.59 |
| 65536 | 93,708.9 | 699.8 | 13.32 |
| 131072 | 312,653.4 (5.2 min) | 419.3 | 11.27 |
| 262144 | 1,218,595.8 (20.3 min) | 215.1 | 8.50 |

All rows: BH Loudbox / P150×8, 1×8 mesh, L36V24, 25 image tokens in the prompt.

**Metric definitions:**
- **ISL** — input sequence length (prompt tokens, including the 25 image tokens)
- **TTFT** — vision replay + one prefill replay (time to first decode logits)
- **Prefill tok/s** — `prompt_len / prefill_trace_replay_time` (compile pass excluded)
- **Decode tok/s/user** — steady-state decode: `decode_iters / decode_trace_replay_total` (compile + capture excluded)

### Running Performance Tests

```bash
# End-to-end (full 36 text + 24 vision layers; also has an L1V1 fast variant)
pytest models/experimental/mistral_small_4_119b/tests/perf/test_e2e_performant.py -m models_performance_bare_metal

# Single-layer device perf
pytest models/experimental/mistral_small_4_119b/tests/perf/test_perf.py -m models_device_performance_bare_metal

# Single-layer prefill+decode device perf (Tracy per-op CSV/JSON; text decoder only, no vision)
pytest models/experimental/mistral_small_4_119b/tests/perf/test_device_perf_single_layer_prefill_decode.py \
    -m models_device_performance_bare_metal
```

`test_device_perf_single_layer_prefill_decode.py` profiles **one text decoder layer** (vision tower
not involved — it builds `TtMistral4TextModel` on text-only weights). It holds two tests:
`test_profile_single_layer_prefill_decode` is the Tracy *workload* (1-layer model, one warmup
iteration, then prefill 128 tokens + one decode step at position 128 inside a `start`/`stop`
signpost window, prefill before decode so per-op tooling can split the phases — no PCC assertion);
`test_device_perf_single_layer_prefill_decode` is the *wrapper* (normal entry point) that runs the
workload under Tracy via `run_device_perf`, then writes `device_perf_*.csv` + a partial benchmark
JSON via `prep_device_perf_report` (no golden perf assertion).

### Repository Layout

```
models/experimental/mistral_small_4_119b/
├── README.md
├── constants.py
├── demo/
│   └── demo_multimodal.py
├── tests/
│   ├── mesh_param.py
│   ├── test_vision_tower_pcc.py
│   ├── test_vision_tower_smoke.py
│   ├── test_multimodal_projector_pcc.py
│   ├── test_multimodal_pcc_unified.py
│   ├── test_text_prefill_pcc.py
│   ├── test_text_prefill_smoke.py
│   ├── test_text_decode_pcc.py
│   ├── test_text_decode_smoke.py
│   ├── test_language_demo_smoke.py
│   └── perf/
│       ├── test_e2e_performant.py
│       ├── test_perf.py
│       └── test_device_perf_single_layer_prefill_decode.py
└── tt/
    ├── mistral3_for_conditional_generation_unified.py   # top-level orchestrator (unified load)
    ├── mistral3_for_conditional_generation_helpers.py
    ├── mistral4_text_model.py                           # text decoder: embed, decode, lm_head
    ├── mistral4_text_prefill.py                         # chunked prefill path
    ├── mistral4_self_attention.py                       # MLA self-attention (latent KV cache)
    ├── mistral4_moe.py                                  # MoE: 128 experts, 4 active + shared
    ├── mistral4_multimodal_projector.py                 # spatial-merge ×2 + GELU
    ├── mistral4_vision_tower.py                         # Pixtral vision tower (24 layers)
    ├── mistral4_vision_attention.py
    ├── mistral4_vision_mlp.py
    ├── mistral4_vision_rope.py                          # vision 2D RoPE
    └── vision_matmul_config.py                          # vision matmul program configs
```

### Known Limitations

| Item | Value |
|------|-------|
| Tested context | **~16K** PCC-verified (vision + language, PCC ≈ 0.85); beyond that the chunked HF CPU reference forward takes prohibitively long — a 64K run was killed after **> 1 h without completing** — so it is the verification bottleneck (not the device) |
| Multi-device | Tensor-parallel sharding across 8 chips — not batch parallelism |
| Prefill tracing | Prefill is chunked and not trace-captured (host token readback per call) |

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/mistral_small_bh_v2_bkp)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/mistral_small_bh_v2_bkp)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ign/mistral_small_bh_v2_bkp)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/mistral_small_bh_v2_bkp)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/mistral_small_bh_v2_bkp)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/mistral_small_bh_v2_bkp)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/mistral_small_bh_v2_bkp)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/mistral_small_bh_v2_bkp)

### CODE REVIEW

**Comment by moritztng** [comment]
One point:

- **Run and document full-depth PCC**. The PCC tests default to 2 text decoder layers (MISTRAL4_PREFILL_N_LAYERS=2) and the decode test validates layer 0 only. The tests are also gated behind environment variables (MISTRAL4_PREFILL_PCC=1) so they skip by default. There is no PCC test that runs the full 36-layer text decoder. DOTS-OCR validates text prefill PCC at full model depth. Run the full 36-layer PCC, document the measured result.

**Comment by ign-ssunil** [comment]
> One point:
> 
> Run and document full-depth PCC**. The PCC tests default to 2 text decoder layers (MISTRAL4_PREFILL_N_LAYERS=2) and the decode test validates layer 0 only. The tests are also gated behind environment variables (MISTRAL4_PREFILL_PCC=1) so they skip by default. There is no PCC test that runs the full 36-layer text decoder. DOTS-OCR validates text prefill PCC at full model depth. Run the full 36-layer PCC, document the measured result.

Addressed. There's now a full 36-layer text-decoder PCC test for both the prefill and decode paths, neither gated behind an env var, and the measured results are documented in the PR description.
